### DISCUSSION TITLE
Hide node element on GLSP editor on click event

### DISCUSSION THREAD

**Discussion by Abhay0294** [eclipse-glsp/glsp#700] [Q&A] [answered]
Hi,

My requirement is as follows:
1. There will be one button on the GLSP editor(ex: Hide button).
2.  On the click of that button the selected node element should get hidden.[Requirement]

Is there any way how can I achieve this ?

**Comment by martin-fleck-at** [comment, marked as answer]
Hi @Abhay0294,

1. There are several ways you could add a button to the GLSP editor, depending on where you would like to trigger the action
   i. You could add a button to the [tool palette](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tool-palette/tool-palette.ts#L175) which triggers an action for all selected elements. You can retrieve the selected elements through the editor context.
   ii. If you use the Theia integration, you could use a context menu entry for Show/Hide, similar to how it is done for [deletion in the workflow example](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/context-menu/delete-element-context-menu.ts).
   iii. You can overlay the diagram with a custom HTML element that is rendered as a [UI Extension](https://www.eclipse.org/glsp/documentation/ui-extensions/).
   iv. Any mouse tool interaction or anything else you can think of ;-)

2. If you only want to hide the element from the user (without any involvement of the server), the easiest way is probably to trigger a [ModifyCSSFeedbackAction](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tool-feedback/css-feedback.ts) for the selected elements and apply or remove a custom CSS class on all selected elements. The CSS class should then properly hide the element, e.g.: `
.invisible {
   fill-opacity: 0.0;
   opacity: 0.0;
   visibility: hidden;
}
` or something like that. If the server needs to know about it, you could create a custom action that has all necessary information.

**Reply by Abhay0294** [reply]
Thanks a lot @martin-fleck-at  for the suggested answer. Just want to ask one more thing, is it possible to hide all the elements on the GLSP editor. For example:
**Scenario 1.** There are 5 node elements and when a user click on the Hide button all the elements should be hidden.
**Scenario 2.** There are 5 node elements with one selected element and when a user click on the Hide button all the elements should be hidden.

**Reply by martin-fleck-at** [reply]
@Abhay0294 Sure, absolutely, the approach is not limited to just the selected nodes. In the case of the ModifyCSSFeedbackAction you just provide a list of element ids on which the classes should be applied. Without testing it, it is probably enough to just provide the root element id or you need to give a list of all element IDs. When the action is executed as a [command](https://github.com/eclipse-glsp/glsp-client/blob/7ff1b00a1575eb1979aa69e8e2873807883d9c6a/packages/client/src/features/tool-feedback/css-feedback.ts#L50) you have access to all elements through the CommandExecutionContext.

**Reply by Abhay0294** [reply]
Hello @martin-fleck-at I have able to hide node elements as per your suggestion.

But there are some abnormilities which I am facing now:
1. I am having two kinds of nodes on the editor namely "TX" & "RX". I have an UI which consists of 2 buttons on editor.
2. When I click on Button A "TX" nodes should get hidden and when clicked on Button B "RX" nodes should get hidden.
3.  Functionality wise it is working fine.
4. when button A is clicked and "TX" nodes are hidden and only "RX" nodes are visible now. But when I select that available "RX" node and start dragging the element on the screen and leave the mouse, all the hidden "TX" nodes again appears.**[This is the abnormal behaviour.]**

Below is the code for hiding the elements.

` getStartViewDataElements(): void {
        const editorCtxData = this.editorContextService.modelRoot.children ? this.editorContextService.modelRoot.children : undefined;
        if (editorCtxData) {
            editorCtxData.map(item => {
                if (item && (item instanceof RectangularNode)) {
                    const node = item as any;
                    if (node.direction === 'RX') { this.startViewData.push(item); }
                }
            });
        }
        this.onHideEditorElements();
    }

    getStopViewDataElements(): void {
        const editorCtxData = this.editorContextService.modelRoot.children ? this.editorContextService.modelRoot.children : undefined;
        if (editorCtxData) {
            editorCtxData.map(item => {
                if (item && (item instanceof RectangularNode)) {
                    const node = item as any;
                    if (node.direction === 'TX') { this.stopViewData.push(item); }
                }
            });
        }
        this.onHideEditorElements();
    }

    onHideEditorElements(): void {
        if (HideElementsAction.buttonID === 'start-button' && this.startViewData.length > 0) {
            for (let i = 0; i < this.startViewData.length; i++) {
                this.feedbackDispatcher.registerFeedback(this, [applyCssClasses(this.startViewData[i], 'invisible')]);
            }
        }
        else if (HideElementsAction.buttonID === 'stop-button' && this.stopViewData.length > 0) {
            for (let i = 0; i < this.stopViewData.length; i++) {
                this.feedbackDispatcher.registerFeedback(this, [applyCssClasses(this.stopViewData[i], 'invisible')]);
            }

        }
    }`

Below id the di.config file:
`export const HideElementsModule = new ContainerModule((bind, _unbind, isBound, rebind) => {
    const context = { bind, _unbind, isBound, rebind };
    configureActionHandler(context, HideElementsAction.KIND, HideElementsActionHandler);
});`

**Can you please suggest me what's going wrong in the implementation.**
<img width="947" alt="GLSP_Issue" src="https://user-images.githubusercontent.com/57217189/179728888-f910a76d-dcd4-479a-b497-9a8bfdd9e12f.PNG">
]

**Reply by martin-fleck-at** [reply]
@Abhay0294 Simply form the code I do not see what could be the issue unfortunately. The general idea is that when you register something as feedback, it will be re-applied even if the model updates (e.g., after ending the drag). When debugging, can you see whether the FeedbackActionDispatcher has the correct feedback registered and calls the ModifyCssFeedbackCommand again after the update?

**Reply by Abhay0294** [reply]
Hello @martin-fleck-at , thanks for your response.
I have tried what you suggested and concluding as below:
1. While dragging or after dragging none of the code line gets hit in css-feedback.ts file during debugging.
2.  After dragging completed invisible nodes again appears.

Requesting you to please look into below reported bug, I think it is sprotty related functionality which resets while dragging/panning.
`https://github.com/eclipse-glsp/glsp/discussions/701`

**Reply by martin-fleck-at** [reply]
@Abhay0294 I am not sure that the bug is related but it seems that the bug in Sprotty was already fixed: https://github.com/eclipse/sprotty/issues/299. So as soon as there is another Sprotty release we can upgrade GLSP as well to see if that resolves your issue.

**Reply by Abhay0294** [reply]
Yes @martin-fleck-at , waiting for that release of sprotty.

**Reply by martin-fleck-at** [reply]
Please note that there is also a PR open on GLSP side for this: https://github.com/eclipse-glsp/glsp-client/pull/191
